app/models/assignment.rb

```diff
@@ -775,13 +779,31 @@ def get_num_assigned(ta_id = nil)
 
   def get_num_marked(ta_id = nil)
     if ta_id.nil?
-      groupings.count(marking_completed: true)
+      groupings.select(&:marking_completed?).count
     else
-      n = 0
-      ta_memberships.includes(grouping: [{current_submission_used: [:submitted_remark, :results]}]).where(user_id: ta_id).find_each do |x|
-        x.grouping.marking_completed? && n += 1
+      if is_criteria_mark?(ta_id)
+        criteria = []
+        criteriatype = []
+        self.criterion_ta_associations.where(ta_id: ta_id).find_each do |x|
+          criteria << x.criterion_id
+          criteriatype << x.criterion_type
+        end
+        n = 0
+        ta_memberships.includes(grouping: :current_submission_used).where(user_id: ta_id).find_each do |x|
+          hasnull = false
+          for i in 0..criteria.size do
```

Comment:
This is a good start. Let's try to remove the need for an inner loop by using a join. (http://guides.rubyonrails.org/active_record_querying.html#joining-tables) I think you need a custom string for the join, something like `"INNER JOIN on results.markable_id = criterion_ta_associations.criterion_id AND results.markable_type = criterion_ta_associations.criterion_type"`.